### PR TITLE
Fix flaky CallbackTests.TestCallbackHandlers Test

### PR DESCRIPTION
**Pull request by suhsteve**
Instead of checking that reading from the socket throws an exception, check `CallbackConnection` to verify if it is running instead.

Cancelling the `CancellationToken` changes the _isRunning flag on the `CallbackConnection` from true to false.  Testing that this flag has been set correctly is the right thing to verify versus checking communication states/exceptions while trying to read the socket streams.

### CODE REVIEW

**Comment by imback82** [comment]
> Instead of checking that reading from the socket throws an exception, check `CallbackConnection` to verify if it is running instead.

Can you update the PR description why the previous behavior could be flakey?

**Comment by suhsteve** [comment]
> > Instead of checking that reading from the socket throws an exception, check `CallbackConnection` to verify if it is running instead.
> 
> Can you update the PR description why the previous behavior could be flakey?

Updated.

**Comment by imback82** [comment]
> checking communication states/exceptions while trying to read the socket streams.

Do you know why this was flaky?

**Comment by suhsteve** [comment]
> > checking communication states/exceptions while trying to read the socket streams.
> 
> Do you know why this was flaky?

No I wasn't able to pinpoint why exactly the socket wasn't throwing an exception.

File: src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs [hunk 1]
```diff
@@ -153,17 +153,33 @@ public void TestCallbackHandlers()
             clientSocket.Connect(ipEndpoint.Address, ipEndpoint.Port);
 
             var callbackConnection = new CallbackConnection(0, clientSocket, callbackHandlersDict);
-            Task.Run(() => callbackConnection.Run(token));
+            Task task = Task.Run(() => callbackConnection.Run(token));
 
-            using ISocketWrapper serverSocket = serverListener.Accept();
-            WriteAndReadTestData(serverSocket, callbackHandler, inputToHandler, token);
+            if (token.IsCancellationRequested)
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:158]: How have we verified this change removes the flakiness of these tests? Do we need to run this multiple times to confirm?

Comment (suhsteve) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:158]: Yes, this has already been verified.

File: src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs [hunk 2]
```diff
@@ -153,17 +153,33 @@ public void TestCallbackHandlers()
...
+                using ISocketWrapper serverSocket = serverListener.Accept();
+                WriteAndReadTestData(serverSocket, callbackHandler, inputToHandler);
+
+                if (callbackHandler.Throws)
+                {
+                    Assert.False(callbackConnection.IsRunning);
+                }
+                else
+                {
+                    Assert.True(callbackConnection.IsRunning);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: Question. When is this running `callbackConnection` stopped?

Comment (suhsteve) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: The callbackConnection can be stopped when the JVM side closes the connection cleanly, with a "CallbackFlags.CLOSE` request, or when reading the stream the socket is closed unexpectedly, or when the cancellation token is triggerd (ie, when the callback server shutsdown).

Comment (imback82) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: I mean in this function?

Comment (suhsteve) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: It'll should stop once the GC cleans up the sockets.

Comment (suhsteve) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: Or if we can manually trigger the cancellation token after each test.

Comment (imback82) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:174]: Yea, can you make the clean up deterministic please?

Comment (suhsteve) [reply src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs:176]: CallbackConnections stopped.

**Review by imback82 — APPROVED** [review]
LGTM, thanks @suhsteve!

**Pull request merged**